content/6_Advanced/SCC.mdx

```diff
@@ -202,11 +202,87 @@ int main() {
 
 <FocusProblem problem="satEx" />
 
-<IncompleteSection>
+Model each topping as a boolean variable and each wish as a clause, so all wishes define a boolean formula in 2-CNF. Then we rewrite each clause $x_i \vee x_j$ as $(\bar x_i \rarr x_j) \wedge (\bar x_j \rarr x_i)$ and construct the implication graph with each variable and its negation as a vertex. For each appearance of $u \rarr v$ in the rewritten formula, an edge from $u$ to $v$ is inserted into the implication graph. (For a diagram, see the explanation at cp-algo linked below.) Then, the formula is satisfiable if and only if no variable lies in the same Strongly Connected Component as its negation, since this would form a contradiction. So we find the SCC of each variable using Kosarajus algorithm. If the formula is satisfiable, we need to pick the value for a variable in the topologically greater component, as otherwise a contradiction could be formed.
 
-implementation
+<Spoiler title="Solution">
 
-</IncompleteSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+size_t const M = 100000;
+
+// x_i lies at index 2 * i, it's negation at index 2 * i + 1.
+vector<unsigned> g[2 * M], rg[2 * M], postorder;
+bool visited[2 * M];
+unsigned scc[2 * M];
+
+void traverse_postorder(unsigned u)
+{
+    visited[u] = 1;
```

Comment:
```suggestion
    visited[u] = true;
```

<del>doesn't really matter but</del> what about using `true` is we are using bool arr anyway?